### PR TITLE
Add HalfFlat distribution

### PR DESCRIPTION
**Pull request by aseyboldt**
I also removed the median from the `Flat` distribution and switched to using `self._default = 0` as default value.

### CODE REVIEW

File: pymc3/distributions/continuous.py [hunk 1]
```diff
@@ -173,8 +172,8 @@ class Flat(Continuous):
     """
 
     def __init__(self, *args, **kwargs):
-        super(Flat, self).__init__(*args, **kwargs)
-        self.median = 0
+        self._default = 0
```
Comment (fonnesbeck) [inline pymc3/distributions/continuous.py:176]: Why the underscore?

Comment (aseyboldt) [reply pymc3/distributions/continuous.py:176]: It's not something anyone outside this class should ever access.

Comment (fonnesbeck) [reply pymc3/distributions/continuous.py:176]: Can we not just populate `self.defaults` directly?

Comment (aseyboldt) [reply pymc3/distributions/continuous.py:176]: Ah, I guess we could just override the `default` method. It doesn't look like anything is using the `defaults` attribute directly (except for one call in `transforms.py`, which looks fine). If it had an underscore we'd have an easier time doing stuff like that ;-)

Comment (aseyboldt) [reply pymc3/distributions/continuous.py:176]: Or did you mean
```python
def __init__(self, *args, **kwargs):
    super()
    self.defaults = ('_default',)
```

**Comment by junpenglao** [comment]
There is conflict. 
Otherwise is it ready?

**Comment by aseyboldt** [comment]
@fonnesbeck is `_default` fine the way it is, or would you rather I changed that somehow?

**Comment by junpenglao** [comment]
I had the same question as @fonnesbeck. For example, we can do:
```python
pm.Normal.dist(0, 1).default()
```
Why not the same for Flat? Is that the same as turning `_default` into `default`?

**Comment by aseyboldt** [comment]
@junpenglao You can still do that:
```python
pm.Flat.dist().default()
array(0.0, dtype=float32)
```
The only difference is that it doesn't store that 0 in the `median` attribute anymore (which I'd argue is wrong), but in a `_default` attribute.

**Comment by junpenglao** [comment]
I see, I think it does make sense to do `_default` then. So basically it is to provide a value to `default` but not calling it `median` as it will not be right.

**Comment by aseyboldt** [comment]
exactly

**Comment by twiecki** [comment]
LGTM.

**Pull request merged**